### ISSUE TITLE
Finetuning Cross-Encoder model results in constant prediction but when validated on test set it shows 100 F1 and accuracy by evaluator function

### ISSUE THREAD

**Issue by ratikapoor** [UKPLab/sentence-transformers#2282]
While using cross-encoder (CrossEncoder('distilroberta-base',num_labels=1)), I'm getting F1 score of 1 on the validation set with labels 0 and 1 after fine tuning. Although fine tuning is not happening properly. While inferencing it gives a constant score of 0.4118. Then also in logs, it shows 100 accuracy and F1 score.
CEBinaryClassificationEvaluator: Evaluating the model on fk-dev dataset after epoch 1:
Accuracy: 100.00 (Threshold: 0.4118)
F1: 100.00 (Threshold: 0.4118)
Precision: 100.00
Recall: 100.00
Average Precision: 30.00

I'm not able to understand why it is not learning anything where I have 11k rows in the data with the target label equally distributed.

**Comment by progsi** [comment]
I am having the same issue. Any news on this? I am using the model like a normal torch model with switching between train and test modes.

**Comment by FDSRashid** [comment]
Me too :(

**Comment by tomaarsen** [comment]
Hello!

I'm trying to reproduce this by running this training script: [training_quora_duplicate_questions.py](https://github.com/UKPLab/sentence-transformers/blob/master/examples/training/cross-encoder/training_quora_duplicate_questions.py), which also uses `CEBinaryClassificationEvaluator` and `distilroberta-base`. (I also limited the training samples to 10000 & increased the batch size to keep the training fast)

And in my case, the accuracies and thresholds do update every epoch:
```
2023-12-21 10:22:05 - Read train dataset
2023-12-21 10:22:05 - Read dev dataset
Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at distilroberta-base and are newly initialized: ['classifier.out_proj.bias', 'classifier.out_proj.weight', 'classifier.dense.weight', 'classifier.dense.bias']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
2023-12-21 10:22:07 - Use pytorch device: cuda
2023-12-21 10:22:07 - Warmup-steps: 32
Iteration: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████| 79/79 [00:34<00:00,  2.26it/s]
2023-12-21 10:22:42 - CEBinaryClassificationEvaluator: Evaluating the model on Quora-dev dataset after epoch 0:                                                
2023-12-21 10:23:04 - Accuracy:           76.20 (Threshold: 0.5443)                                                                                            
2023-12-21 10:23:04 - F1:                 70.19 (Threshold: 0.2527)
2023-12-21 10:23:04 - Precision:          59.36
2023-12-21 10:23:04 - Recall:             85.86
2023-12-21 10:23:04 - Average Precision:  69.16

2023-12-21 10:23:04 - Save model to output/training_quora-2023-12-21_10-22-05
Iteration: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████| 79/79 [00:54<00:00,  1.44it/s] 
2023-12-21 10:23:59 - CEBinaryClassificationEvaluator: Evaluating the model on Quora-dev dataset after epoch 1:                                                
2023-12-21 10:24:22 - Accuracy:           79.98 (Threshold: 0.5933)                                                                                            
2023-12-21 10:24:22 - F1:                 73.66 (Threshold: 0.3897)
2023-12-21 10:24:22 - Precision:          64.85
2023-12-21 10:24:22 - Recall:             85.25
2023-12-21 10:24:22 - Average Precision:  74.99

2023-12-21 10:24:22 - Save model to output/training_quora-2023-12-21_10-22-05
Iteration: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████| 79/79 [00:53<00:00,  1.47it/s] 
2023-12-21 10:25:16 - CEBinaryClassificationEvaluator: Evaluating the model on Quora-dev dataset after epoch 2:                                                
2023-12-21 10:25:38 - Accuracy:           80.21 (Threshold: 0.5649)                                                                                            
2023-12-21 10:25:38 - F1:                 73.90 (Threshold: 0.2913)
2023-12-21 10:25:38 - Precision:          65.38
2023-12-21 10:25:38 - Recall:             84.98
2023-12-21 10:25:38 - Average Precision:  75.69

2023-12-21 10:25:38 - Save model to output/training_quora-2023-12-21_10-22-05
Iteration: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████| 79/79 [00:55<00:00,  1.41it/s] 
2023-12-21 10:26:35 - CEBinaryClassificationEvaluator: Evaluating the model on Quora-dev dataset after epoch 3:                                                
2023-12-21 10:26:57 - Accuracy:           80.28 (Threshold: 0.6295)                                                                                            
2023-12-21 10:26:57 - F1:                 73.97 (Threshold: 0.3453)
2023-12-21 10:26:57 - Precision:          65.85
2023-12-21 10:26:57 - Recall:             84.38
2023-12-21 10:26:57 - Average Precision:  75.93

2023-12-21 10:26:57 - Save model to output/training_quora-2023-12-21_10-22-05
Epoch: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 4/4 [04:50<00:00, 72.58s/it]
```

Perhaps you could provide a bit more information about what differs between your case & this training example? It might allow us to figure out what is wrong.

- Tom Aarsen

**Comment by tomaarsen** [comment]
Hello!

Thanks for the extra info. I'll try to do some experimentation as well. As a quick heads up, `InputExample` from `sentence_transformers.readers` and from `sentence_transformers` are the same, so no need to worry about that.

- Tom Aarsen

**Comment by FDSRashid** [comment]
Thank you for checking it! The task is very similar to the quora script: between two short texts, determine if they are similar or disimilar. Dissimilar pairs of texts have different content. I know the dataset prep works because I trained a Bi Encoder using the same eval task, now I'm training the cross encoder. My entire dataset has two columns : a text and an integer grouping label to indicate similarity. I took pairs of texts from the same group and pairs of texts from different groups for training the cross encoder model. Hopefully this might add context?

**Comment by HaokaiDing** [comment]
I can reproduce the evaluator part on current `main` without training or downloading a model:

```python
import numpy as np
from sentence_transformers.sentence_transformer.evaluation import BinaryClassificationEvaluator

scores = np.array([0.5, 0.5])
labels = np.array([1, 0])
print(BinaryClassificationEvaluator.find_best_acc_and_threshold(scores, labels, True))
print(BinaryClassificationEvaluator.find_best_f1_and_threshold(scores, labels, True))
```

The current helpers report accuracy 1.0 and F1 1.0 at threshold 0.5. Applying that threshold predicts both pairs positive, which gives accuracy 0.5 and F1 2/3.

The threshold search evaluates splits between individual rows even when their scores are equal, so it can report a partition no threshold can produce. It also omits the all-positive/all-negative endpoints.

I have prepared a fix that evaluates complete score groups and includes the endpoints. The new tests compare the returned thresholds and metrics against sklearn, cover both score directions and tied-label orders, and keep the existing ordinary distance threshold behavior. On the original code, 18 of the 20 parameterized cases fail. With the fix, 29 model-free tests pass, including the existing tests.

This addresses the incorrect evaluator metrics; I have not reproduced or diagnosed why the original fine-tuned models produce constant predictions.

Prepared with OpenAI Codex; the reproduction and tests were run locally.
